Document how to restore etcd on OCP4 (#120)
appuio/openshift4-docs: docs/modules/ROOT/partials/recovery/recover-etcd.adoc

[abstract]
--
ifeval::["{provider}" == "cloudscale"]
Steps to recover etcd on an OpenShift 4 cluster on https://www.cloudscale.ch/[cloudscale.ch].
endif::[]
ifeval::["{provider}" == "exoscale"]
Steps to recover etcd on an OpenShift 4 cluster on https://www.exoscale.com/[Exoscale].
endif::[]
--

[WARNING]
====

> Restoring to a previous cluster state is a destructive and destablizing action to take on a running cluster. This should only be used as a last resort.

Reviewer: ```suggestion
Restoring to a previous cluster state is a destructive and destabilizing action to take on a running cluster.
This should only be used as a last resort.
```

If you are able to retrieve data using the Kubernetes API server, then etcd is available and you shouldn't restore using an etcd backup.
====

== Starting situation

ifeval::["{provider}" == "cloudscale"]
* You have an OpenShift 4 cluster on cloudscale.ch
endif::[]
ifeval::["{provider}" == "exoscale"]
* You have an OpenShift 4 cluster on Exoscale
endif::[]
* One of the following scenarios is true:

> ** The cluster has lost the majority of control plane hosts (quorum loss).

Reviewer: Maybe we should add something here that this procedure won't work if we've lost all control plane hosts. Not sure how to formulate it though.

```suggestion
** The cluster has lost the majority of its control plane hosts (quorum loss).
```

> ** An administrator has deleted something critical and must restore to recover the cluster.

Reviewer: In this scenario, the cluster object backup should be sufficient in most cases, I think.

```suggestion
** An administrator has deleted a critical component which cannot be restored from the object backup.
```

export RESTIC_REPOSITORY="s3:${RESTIC_ENDPOINT}/${RESTIC_BUCKET}"
echo $RESTIC_REPOSITORY
rm -rf cluster-repo
----

include::partial$connect-to-vault.adoc[]

.Fetch backup secrets from vault

[source,bash]
----
export RESTIC_PASSWORD=$(vault kv get \
  -format=json "clusters/kv/${TENANT_ID}/${CLUSTER_ID}/cluster-backup" | jq -r '.data.data.password')
ifeval::["{provider}" == "cloudscale"]
export AWS_ACCESS_KEY_ID=$(vault kv get \
  -format=json "clusters/kv/${TENANT_ID}/${CLUSTER_ID}/cloudscale" | jq -r '.data.data.s3_access_key')
export AWS_SECRET_ACCESS_KEY=$(vault kv get \
  -format=json "clusters/kv/${TENANT_ID}/${CLUSTER_ID}/cloudscale" | jq -r '.data.data.s3_secret_key')
endif::[]
ifeval::["{provider}" == "exoscale"]
export AWS_ACCESS_KEY_ID=$(vault kv get \
  -format=json "clusters/kv/${TENANT_ID}/${CLUSTER_ID}/exoscale/storage_iam" | jq -r '.data.data.s3_access_key')
export AWS_SECRET_ACCESS_KEY=$(vault kv get \
  -format=json "clusters/kv/${TENANT_ID}/${CLUSTER_ID}/exoscale/storage_iam" | jq -r '.data.data.s3_secret_key')
endif::[]
----

> .Download files from latest etcd snapshot
> 
> [source,bash]
> ----
> SNAPSHOT_ID=$(restic snapshots --json --latest=1 --path /syn-cluster-backup-etcd-etcd-backup.tar.gz | jq -r '.[0].id')
> restic dump "${SNAPSHOT_ID}" /syn-cluster-backup-etcd-etcd-backup.tar.gz | tar xzv
> ----

Reviewer: Might be worth it to create a temporary directory to download the backup to?

```suggestion
.Download files from latest etcd snapshot

[source,bash]
----
TEMP_DIR=$(mktemp -d)
pushd ${TEMP_DIR}
SNAPSHOT_ID=$(restic snapshots --json --latest=1 --path /syn-cluster-backup-etcd-etcd-backup.tar.gz | jq -r '.[0].id')
restic dump "${SNAPSHOT_ID}" /syn-cluster-backup-etcd-etcd-backup.tar.gz | tar xzv
----
```

== Connect to Master Node by SSH

.Fetch the ssh key

[source,bash,subs="attributes+"]
----
vault kv get -format=json clusters/kv/${TENANT_ID}/${CLUSTER_ID}/{provider}/ssh \
  | jq -r '.data.data.private_key' | base64 --decode > ssh_key
chmod 400 ssh_key
----

NOTE: The following steps are VSHN specific

.Find load balancer host

[source,bash]
----
LB_HOST=$(ssh management1.corp.vshn.net grep -E "^Host.*${CLUSTER_ID}" '~/.ssh/config' | head -1 | awk '{print $2}')
echo $LB_HOST
----

> NOTE: Ensure your ssh config is up-to-date: `ssh management1.corp.vshn.net sshop generate -li`.

Reviewer: Generally, engineers will have the configuration generated by `sshop` on their local machine. This indirection to `management1.corp.vshn.net` seems rather unnecessary.

Author: I wasn't sure how common `sshop` configuration is on local machines. 👍

.Upload recovery files to master node

[source,bash]
----
MASTER_NODE=etcd-0

> scp -J management1.corp.vshn.net,"${LB_HOST}" -i ssh_key static_kuberesources_*.tar.gz snapshot_*.db "core@${MASTER_NODE}:"

Reviewer: Same here, usually it's not necessary to use `management1.corp.vshn.net` as jumphost.

----

.Connect to master node

[source,bash]
----

> ssh -J management1.corp.vshn.net,"${LB_HOST}" -i ssh_key "core@${MASTER_NODE}"

Reviewer: And again here, the jumphost should not be required on a properly setup system.
